agent/engine/docker_client.go

```diff
@@ -1038,6 +1043,99 @@ func (dg *dockerGoClient) createVolume(ctx context.Context,
 	return volumeResponse{Volume: volume, Error: nil}
 }
 
+func (dg *dockerGoClient) InspectVolume(name string, timeout time.Duration) volumeResponse {
+	// Create a context that times out after the 'timeout' duration
+	// Injecting the 'timeout' makes it easier to write tests.
+	// Eventually, the context should be initialized from a parent root context
+	// instead of TODO.
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	// Buffered channel so in the case of timeout it takes one write, never gets
+	// read, and can still be GC'd
+	response := make(chan volumeResponse, 1)
+	go func() { response <- dg.inspectVolume(ctx, name) }()
+
+	// Wait until we get a response or for the 'done' context channel
+	select {
+	case resp := <-response:
+		return resp
+	case <-ctx.Done():
+		// Context has either expired or canceled. If it has timed out,
+		// send back the DockerTimeoutError
+		err := ctx.Err()
+		if err == context.DeadlineExceeded {
+			return volumeResponse{Volume: nil, Error: &DockerTimeoutError{timeout, "inspecting"}}
+		}
+		// Context was canceled even though there was no timeout. Send
+		// back an error.
+		return volumeResponse{Volume: nil, Error: &CannotInspectVolumeError{err}}
+	}
+}
+
+func (dg *dockerGoClient) inspectVolume(ctx context.Context, name string) volumeResponse {
+	client, err := dg.dockerClient()
+	if err != nil {
+		return volumeResponse{Volume: nil, Error: CannotGetDockerClientError{version: dg.version, err: err}}
+	}
+
+	dockerVolume, err := client.InspectVolume(name)
+	if err != nil {
+		return volumeResponse{Volume: nil, Error: CannotInspectVolumeError{err}}
+	}
+
+	volume := taskresource.NewVolumeResource(
+		dockerVolume.Name,
+		dockerVolume.Mountpoint,
+		dockerVolume.Driver,
+		dockerVolume.Labels)
+	return volumeResponse{Volume: volume, Error: nil}
+}
+
+func (dg *dockerGoClient) RemoveVolume(name string, timeout time.Duration) error {
+	// Create a context that times out after the 'timeout' duration
+	// Injecting the 'timeout' makes it easier to write tests.
+	// Eventually, the context should be initialized from a parent root context
+	// instead of TODO.
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	// Buffered channel so in the case of timeout it takes one write, never gets
+	// read, and can still be GC'd
+	response := make(chan error, 1)
+	go func() { response <- dg.removeVolume(ctx, name) }()
+
+	// Wait until we get a response or for the 'done' context channel
+	select {
+	case resp := <-response:
+		return resp
+	case <-ctx.Done():
+		// Context has either expired or canceled. If it has timed out,
+		// send back the DockerTimeoutError
+		err := ctx.Err()
+		if err == context.DeadlineExceeded {
+			return &DockerTimeoutError{timeout, "removing"}
```

Comment:
I think this should be "removing volume".  Please make sure that that's done for "creating volume" as well